ffi/src/lib.rs

```diff
@@ -179,8 +189,12 @@ mod private {
         fn from(val: Vec<bool>) -> Self {
             let len = val.len();
             let boxed = val.into_boxed_slice();
-            let ptr = Box::into_raw(boxed).cast();
-            KernelBoolSlice { ptr, len }
+            let ptr = Box::leak(boxed).as_mut_ptr();
+            if let Some(ptr) = NonNull::new(ptr) {
+                KernelBoolSlice { ptr, len }
+            } else {
+                KernelBoolSlice::empty()
```

Comment:
I don't think the leaked pointer can ever be null, so None case here is impossible?

Reply:
No I don't believe so, I am just pleasing the compiler here because internally a vec that hasn't allocated also won't produce a null here and I can't think of any situation in which it would realistically be null.

Reply:
Looks like there's a suitable [From](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#impl-From%3C%26mut+T%3E-for-NonNull%3CT%3E):
```rust
let ptr = Box::leak(boxed).into();
KernelBoolSlice { ptr, len }
```

Reply:
I was hoping this would work, but we call `.as_mut_ptr()` to get the pointer to the inner buffer of the vec, `NonNull::from()` wants a `&mut T` which I am not aware of anyway to get a mutable _reference_ to the inner buffer (only a pointer) that doesn't involve dereferencing the pointer above, which would add unsafe to the call. I would prefer to keep the check above in lieu of adding unsafe to the method, thoughts?

Reply:
Also keep in mind a lot of ways to refer to the vec in rust give pointers to the rust object of the vec, not the inner buffer, making this harder to get

Reply:
Oops! I had indeed forgotten the `as_mut_ptr` was a vec method (wrongly assumed it was just turning the leaked reference into a pointer). Agree there doesn't seem to be any good way to directly extract a NonNull from a slice, even tho it's guaranteed to exist (grr). All the examples just just `expect` to unwrap the option when passing a reference to `NonNull::new`, should we do that?

Reply:
Another possibility would be to leverage [first_mut](https://doc.rust-lang.org/std/primitive.slice.html#method.first_mut), but that's still annoying and ugly because it returns an `Option<&mut T>` (because it's UB to create a reference from a zero-length slice). So `expect` or the current code are probably the best we can do, barring a rust API change.

Reply:
Expect should be okay because if it ever hit that error, we have a legitimate bug we should investigate.